[1단계 - 지하철 노선도 미션] 엘라(김현주) 미션 제출합니다.

bucketHaneul opened the pull request:
꺄하 안녕하세요 제이비님😍 이번에 @zigsong과 지하철 미션을 진행한 엘라입니다! 우테코 들어와서 첫 미션 리뷰을 제이비님께 받았었는데요!!!! 마지막 미션에서 제이비님을 다시 만나게 되었네용 ㅎㅎㅎㅎ 페어를 통해 처음으로 TS로 코드를 작성해보았습니다. 잘 부탁 드리겠습니다 >_<


데모 페이지 https://react-subway-map.firebaseapp.com/

## 🙌 Step1 기능 구현 목록
### 회원 기능
- [x] 사용자는 회원 가입을 할 수 있다.
- [x] 사용자는 가입한 계정으로 로그인을 할 수 있다.
- [x] 로그인한 사용자만 지하철 노선도 관리 페이지에 접근할 수 있다.

### 역 관리 기능

- [x] 사용자는 지하철 역을 추가할 수 있다.
- [x] 사용자는 지하철 역을 삭제할 수 있다.
- [x] 사용자는 지하철 역을 수정할 수 있다.
- [x] 사용자는 등록되어 있는 전체 지하철 역 목록을 조회할 수 있다.

### 노선 관리 기능

- [x] 사용자는 지하철 노선을 추가할 수 있다.
- [x] 사용자는 등록되어 있는 지하철 노선을 삭제할 수 있다.
- [x] 사용자는 등록되어 있는 지하철 노선을 수정할 수 있다.
- [x] 사용자는 등록되어 있는 전체 지하철 노선 목록을 조회할 수 있다.

### 구간 관리 기능

- [x] 사용자는 특정 지하철 노선에 구간을 추가할 수 있다.
- [x] 사용자는 노선에 등록되어 있는 구간을 삭제할 수 있다.
- [x] 사용자는 특정 노선의 전체 구간 목록을 조회할 수 있다.

## 🎈 집중적으로 리뷰받고 싶은 부분
- 비동기 통신에 대한 구조가 잘 잡혀있는지
- 기존 custom Hook에서 개선할 점이 있는지
- custom Hook으로 더 분리할 수 있는 부분이 있는지
- 컴포넌트가 재사용성 있게 잘 구현되었는지

마구마구 피드백 해주세요 제이비님 >_<

woowapark committed:
Update README.md

zigsong committed:
chore: 초기 환경 세팅
- eslint, prettier 설정
- storybook 설치
- emotion 설치

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
docs: 구현 기능 목록 작성

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 페이지 생성 및 라우팅

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: App style 설정 및 BaseLayout 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: NavBar 컴포넌트 구현 및 라우터 연결

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
feat: CardLayout 컴포넌트 구현

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

bucketHaneul committed:
feat: Logo에 라우터 연결

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

bucketHaneul committed:
feat: TextButton 컴포넌트 구현

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

bucketHaneul committed:
feat: TextButton 컴포넌트 구현
Co-authored-by: jieun song <zigsong@users.noreply.github.com>

zigsong committed:
feat: IconButton 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
chore: babel-plugin 설정

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: Input 컴포넌트 구현

zigsong committed:
feat: Dropdown 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: Modal 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: StationPage 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: LinePage 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: SectionPage 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: SectionModal 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: LineModal 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: StationPage 내 역 수정 form 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: LineModal 수정 폼 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: SectionModal에 선택한 노선 전달

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

zigsong committed:
refactor: storybook에 필요한 arguments 전달

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: LineModal, SectionModal의 storybook 작성

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: LoginPage 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: SignupPage 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: useFetch 함수 구현

- 지하철역 불러오기
- 지하철역 생성
- 지하철역 삭제

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

zigsong committed:
feat: 노선 관리 api 적용

- 노선 생성
- 노선 삭제

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 노선 수정 api 적용

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 구간 관리 api 적용

- 구간 생성
- 구간 삭제

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 회원가입 api 적용

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 로그인 api 적용

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
refactor: 노선 정보 최신 상태 반영 로직 개선

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: Notification 컴포넌트 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 역 관리 페이지 input 유효성 검사

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 노선 관리 페이지 input 유효성 검사

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 구간 관리 페이지 input 유효성 검사

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 회원가입 페이지 비밀번호 확인 input 유효성 검사

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 로그인 성공 여부 검사

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
refactor: api 요청 방식 변경

- useFetch 삭제

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: redux toolkit으로 user 상태 전역 관리 적용

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: user의 로그인 여부에 따른 페이지 관리

- 로그인, 회원가입, 로그아웃 기능 구현
- 접근 권한이 없는 페이지 접속 시 랜딩 페이지로 리다이렉션

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: serverSlice  구현

- request에 BASE_URL 전달

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: ServerSelector 컴포넌트 구현

- 로그인/회원가입 시 서버 선택 유도
- 로그아웃 시 선택 서버 초기화

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 컴포넌트에 전달하는 prop interface 이름 통일

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: ServerSelector 스토리북 작성

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
refactor: LineModal, SectionModal 스토리북 인자 전달

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
fix: 새로고침 시 user 로그인 정보 유지

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 유저 인터랙션에 따른 응답 메시지를 스낵바로 표시

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

zigsong committed:
refactor: alert를 snackbar로 변경

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
refactor: modal 리팩토링

- modal open 시 form reset
- 숫자 입력 폼 max값 제한
- dropdown에 value 추가

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 지하철역 이름 수정 기능 구현

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
feat: 회원가입 시 이메일 중복 여부 확인

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

zigsong committed:
feat: 회원가입, 로그인 폼 아이콘 추가

- 서버 선택지 위치 변경

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
fix: end point 수정

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

zigsong committed:
fix: 지하철역 이름 변경 시 유효성 검사

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

bucketHaneul committed:
refactor: ServerSelector 위치 수정

- NavBar margin 조정

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

bucketHaneul committed:
refactor: ServerSelector 위치 수정

-NavBar amrgin 조정

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

zigsong committed:
Merge branch 'step1' of https://github.com/HyuuunjuKim/react-subway-map into zig-step1

bucketHaneul committed:
refactor: station 목록 최신순으로 정렬

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

bucketHaneul committed:
fix: 에러 메세지 없을 경우 예외처리

Co-authored-by: jieun song <zigsong@users.noreply.github.com>

zigsong committed:
feat: 랜딩 페이지 UI 개선

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
refactor: useFetch hook 구현

- api 요청 중복 코드 개선
- Loading 시 spinner 표시

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

zigsong committed:
chore: React Testing Library 초기 설정

Co-authored-by: HyuuunjuKim <HyuuunjuKim@users.noreply.github.com>

bucketHaneul committed:
Merge branch 'zig-step1' of https://github.com/zigsong/react-subway-map into step1

bucketHaneul committed:
chore: firebase 세팅

bucketHaneul committed:
test: RTL로 StationPage 테스팅

bucketHaneul committed:
refactor: constants 내 파일들 모아서 export하는 index.ts 생성

bucketHaneul committed:
refactor: 상수화

- ThemeProvider 삭제
- 각각의 input 별 상수화

bucketHaneul committed:
refactor: useNotify 구현 및 적용

bucketHaneul committed:
docs: 구현한 기능 요구사항 목록에 표시

bucketHaneul committed:
refactor: useNotify에 Notification 컴포넌트 추가

jbee37142 commented on src/components/shared/TextButton/TextButton.tsx:
```diff
@@ -0,0 +1,32 @@
+  onClick,
+}: TextButtonProps) => {
+  return (
+    <Styled.Container
```
이 태그가 `button`이 맞는지 `*.styles.ts`를 다시 봤어야 하는 번거로움이 있었어요 👀

jbee37142 commented on src/components/shared/TextButton/TextButton.tsx:
```diff
@@ -0,0 +1,32 @@
+  onClick?: () => void;
+}
+
+const TextButton = ({
+  text,
+  styleType,
+  sizeType = ButtonSize.SMALL,
+  isSelected,
+  onClick,
```
이 `TextButton` 컴포넌트는 `disabled`처리를 하고 싶을 때 수정이 발생할 것 같아요.
button의 기본 attributes를 받아서 전부 넘겨주는 처리가 있으면 그 사용성이 높아질 수 있을 것 같아요!

bucketHaneul replied:
앗 이 프로젝트에서는 아직 안 쓰는 속성이라는 생각에, 버튼의 기본 속성들에 대한 처리가 부족했네요 ❗ 기본 속성에 대한 props 보완하였습니다 ㅎㅎ

감사합니다 ~!

jbee37142 commented on src/components/shared/TextButton/TextButton.tsx:
```diff
@@ -0,0 +1,32 @@
+      isSelected={isSelected}
+      onClick={onClick}
+    >
+      {text}
```
text라는 별도 props를 둔 것과 그마저 type이 string만 받는 것이 아쉬운데요, (number는요?ㅠㅠ)
children을 통한 합성을 하지 않았던 이유가 궁금합니다! 그리고 왜 TextButton인지? Button으로는 안됐는지도 궁금하게 만드는 props 디자인이었습니다 ㅎㅎ

bucketHaneul replied:
제이비님 말대로, text라는 props로 버튼 내에 들어간 text를 받게 되면 string type의 value만 받게 되네요! 이 점을 인지하지 못했네요 😥😥😥 그래서 text라는 props를 없애고 children으로 버튼 내에 들어갈 text를 받도록 수정했습니다!

그런데 그러다보니 IconButton과 TextButton을 굳이 구분하지 않아도 될 것 같아서, Button이라는 컴포넌트로 합칠 수 있었습니다!!!

감사합니다 😎

jbee37142 commented on src/components/shared/TextButton/TextButton.styles.ts:
```diff
@@ -0,0 +1,60 @@
+};
+
+const buttonSize = {
+  [ButtonSize.LARGE]: css`
+    width: 100%;
+  `,
```
이게 사이즈일지 궁금해요!
기본적으로 button은 width: 100%가 아닌데요, sizeType을 Large로 했더니 width: 100%가 된다?
사이즈를 조정한다면 font-size 또는 height, width가 적당히 조절될 것으로 기대할 것 같습니다!ㅎㅎ

bucketHaneul replied:
앗 동의합니다!!! 사실 Input 컴포넌트도 width만 늘리고 싶을 때는, 사용하는 page 단에서 조정 해주었는데 Button을 왜 이렇게 했을까요 제가 ㅜㅜㅜ Button 컴포넌트를 사용하는 쪽에서 width만 늘려주도록 수정했습니다!

제이비님! 여기서 질문있습니다 🙌 ㅎ ㅎ
혹시 저 ButtonSize에 FULL이라는 type으로 `width: 100%`를 주거나, 혹은 ButtonWidth라는 type을 새로 정의해서 `width: 
 100%`를 가지는 새로운 타입을 정의하는 방법은 어떨까요?? 이번 어플리케이션에서는 사실 ButtonSize가 거의 width 조정만을 위한 type인 것 같아서요!!

jbee37142 replied:
> 혹시 저 ButtonSize에 FULL이라는 type으로 width: 100%를 주거나,

동일한 네이밍의 props에서 값마다 다른 속성이 지정된다면 혼란이 발생할 것 같아요!

> 혹은 ButtonWidth라는 type을 새로 정의해서 width: 100%를 가지는 새로운 타입을 정의하는 방법은 어떨까요?? 이번 어플리케이션에서는 사실 ButtonSize가 거의 width 조정만을 위한 type인 것 같아서요!!

실제로 많이 사용하는 방법입니다!

bucketHaneul replied:
오 그렇군요 ~! 두버째 방법 한 번 시도해봐야겠습니다 감사합니다 ㅎㅎㅎ

jbee37142 commented on src/components/shared/Modal/Modal.tsx:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <Styled.Dimmer isOpen={isOpen} onClick={clickDimmer}>
+      <Styled.Container>
+        <Styled.CloseButton onClick={onClose}>
```
1. close button없이 사용하고 싶을 때는 어떻게 할 수 있을까요?
2. 이 svg말고 다른 아이콘으로 close button으로 modal 컴포넌트를 재사용하고 싶을 땐 어떻게 할 수 있을까요?
3. `title` 없이 사용하고 싶다면 이 컴포넌트는 사용할 수 없을 것 같아요!

bucketHaneul replied:
1,2 -> close button을 선택적으로 받도록 수정하였습니다.
3 -> title을 선택적으로 받도록 수정하였습니다.

Modal Component에 대한 수정 내용을 바탕으로 storybook에 case 추가했습니다.

감사합니다 >_<

jbee37142 commented on src/components/shared/Modal/Modal.tsx:
```diff
@@ -0,0 +1,31 @@
+}
+
+const Modal = ({ isOpen, title, children, onClose }: ModalProps) => {
+  const clickDimmer = (event: React.MouseEvent<HTMLDivElement>) => {
```
컴포넌트 내부에서 정의하는 이벤트 핸들러 네이밍 convention이 있어도 좋겠다는 생각이 드네요 ㅎㅎ
handle-* prefix 같은 것으로 구분을 해봐도 가독성이 좋아질 것 같습니다.

bucketHaneul replied:
앗 인정합니다!

보통 `on-*`이라고 prefix를 붙여서 다음과 같이 수정했습니다! 

```js
const onClickDimmer = (event: React.MouseEvent<HTMLDivElement>) => {
  if (event.target === event.currentTarget) onClose();
};
```

jbee37142 commented on src/components/shared/Modal/Modal.tsx:
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import Styled from './Modal.styles';
+
+interface ModalProps {
```
이 interface는 외부로 export되지 않습니다. (되도록 export하지 않는 것이 맞구요)
그렇다면 네이밍에서 Modal임을 들고 있어야 할까요? `Props`라고만 해도 충분하지 않을까요?

bucketHaneul replied:
으악!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

전체적으로 외부로 export 되지않는 interface를 Props로 네이밍 수정하였는데 Modal 컴포넌트를 놓치고 수정하지 않았나보네요ㅜㅜㅜㅜㅜㅜㅜㅜ흐억 ㅜㅜㅜ 수정하였습니다! 꼼꼼히 봐주셔서 감사합니다 😁

jbee37142 commented on src/components/shared/Modal/Modal.styles.ts:
```diff
@@ -0,0 +1,63 @@
+  align-items: center;
+`;
+
+const CloseButton = styled.div`
```
CloseButton이라면 button element이어야만 합니다 ㅎㅎ

bucketHaneul replied:
반영하였습니다!

이번에 마감기간에 신경쓰느라 전체적으로 제가 디테일한 부분을 많이 놓친 것 같네요,, ㅜ ㅜ 반성해야겠어요

jbee37142 commented on src/components/shared/Modal/Modal.styles.ts:
```diff
@@ -0,0 +1,63 @@
+import styled from '@emotion/styled';
+
+const Dimmer = styled.div<{ isOpen: boolean }>`
+  opacity: ${({ isOpen }) => (isOpen ? 1 : 0)};
+  visibility: ${({ isOpen }) => (isOpen ? 'visible' : 'hidden')};
```
display로 제어하지 않고 visibility로 제어하는 이유가 궁금합니다.

bucketHaneul replied:
처음에는 display로 모달 open을 제어했습니다😁

그런데 조금 더 모달이 부드럽게 열리는 효과를 주고 싶어서 visibilisty와 transition으로 모달을 제어하도록 변경하였습니다!!!

jbee37142 replied:
> 부드럽게 열리는 효과를 주고 싶어서

네! 왠지 그러셨을 것 같아요! 과연 이 목적을 위해 이 방법을 적용하는 것이 맞을지 문제점은 없을지 고민해보면 좋을 것 같습니다~

jbee37142 commented on src/components/shared/Modal/Modal.styles.ts:
```diff
@@ -0,0 +1,63 @@
+const Container = styled.div`
+  background-color: #fdfdfd;
+  border-radius: 5px;
+  position: absolute;
+  top: 50%;
+  left: 50%;
+  transform: translate(-50%, -50%);
```
가운데 정렬 때문인것 같은데 이 style이 필요하던가요?

bucketHaneul replied:
저 style이 없으면 modal이 위로 붙습니다!

![image](https://user-images.githubusercontent.com/43339385/120438646-9837e500-c3bc-11eb-8e69-06bf73f160e6.png)

혹시 다른 방법이 있을까요??? 😮😮😮

jbee37142 replied:
flex로도 중앙 정렬이 가능할텐데요?! 잘 안됐다면 height 문제일 것 같아요

jbee37142 commented on src/components/shared/Modal/Modal.styles.ts:
```diff
@@ -0,0 +1,63 @@
+  height: 100%;
+  background-color: rgba(0, 0, 0, 0.6);
+  transition: opacity 0.25s ease;
+  z-index: 999;
```
z index가 관리되지 않는 위험성이 있을 것 같습니다.

bucketHaneul replied:
제이비님 혹시 이런 z-index는 어떻게 관리해야할까요??
z-index를 제대로 사용해본 적이 없다보니 1부터 차근차근 올리면서 사용해야 할지, 다른 곳에 적어두면서 사용해야할지 살짝 감이 오지 않습니다 ㅠ ㅠ

혹시 제이비님 만의 z-index 관리법이 있을까용?

jbee37142 replied:
정답은 없는 것 같고, 프로덕트 레벨에선 되도록 사용을 지양하면서 context에 따라 관리를 하면 좋겠다 정도네요!

jbee37142 commented on src/components/shared/Modal/Modal.styles.ts:
```diff
@@ -0,0 +1,63 @@
+  padding: 3.2rem 2.8rem;
+  box-sizing: border-box;
+  z-index: 1;
+  border: 1px solid #dddddd;
```
상수화되지 않은 컬러코드 사용으로 추후 유지보수하는데 어려움이 있을 것 같습니다!

bucketHaneul replied:
상수화된 컬러로 사용하도록 변경하였습니다~!

jbee37142 commented on src/components/shared/Modal/Modal.tsx:
```diff
@@ -0,0 +1,31 @@
+  onClose: () => void;
+}
+
+const Modal = ({ isOpen, title, children, onClose }: ModalProps) => {
```
외부에서 width가 조금 다른 모달을 만들고 싶어지는 경우가 생긴다면 이 Modal를 사용하지 못할 것 같아요.
emotion을 사용하고 계시던데 className을 비롯한 다른 최상위 element의 attributes를 전부 전달해주면 어떨까요?

bucketHaneul replied:
죄송하지만 다른 최상위 element의 attributes를 전부 전달해준다는 부분이 `...props`를 통해서 전부 내려주는 방법을 말씀하시는 걸까요??😯😯😯 

제이비님 말씀대로 위 모달과 조금 다른 모달을 사용하고 싶으면 위 모달은 사용하기 힘들 것 같습니다 😯
그런데 여기서 질문이 있습니다! 혹시 `...props`로 다른 attributes를 모두 전달하는 방법이 위험한 방법은 아닌가요?? 저번 프로젝트에서 `eslint-config-airbnb`를 사용할때는 props를 spread하지 못하게 하는 rule이 있어서, 이렇게 전부 다 내려주는게 위험한 방법인가? 하고 혼자 생각해 보았던 기억이 있어서요!

jbee37142 replied:
> 그런데 여기서 질문이 있습니다! 혹시 ...props로 다른 attributes를 모두 전달하는 방법이 위험한 방법은 아닌가요?? 저번 프로젝트에서 eslint-config-airbnb를 사용할때는 props를 spread하지 못하게 하는 rule이 있어서, 이렇게 전부 다 내려주는게 위험한 방법인가? 하고 혼자 생각해 보았던 기억이 있어서요!

rule이 있다는 것은 아는데 상황에 따라 (추상화 레벨에 따라) 다르게 적용해야 한다고 생각해요 ㅎㅎ
추상화 레벨이 낮아 재사용을 높여야 하는 컴포넌트에서는 당연한 방법이라고 생각합니당.

위험하다고는 한번도 생각해본적이 없는데 어떤 점이 위험하게 할까요???? 👀

bucketHaneul replied:
ㅎㅎㅎ 사실,,,,,저번 프로젝트에서 `...props`로 props를 전부 내려주면서 동시에 styled-component로 props를 다시 전부 내리는 바람에,,,,,개발자 도구를 보니 onClick, type같은 부분이 css단 까지 넘어가서, 전부 선으로 찍 그어진 상태로 개발자 도구 Styles에 존재하는 장면을 목격했습니다 ㅜ ㅜ

물론 위 케이스는 저의 실수였고 지금과는 상황이 다르지만,,,그 때 이후로 ...props가 살짝(?) 두려워지더라구요 ㅎㅎㅎ 호호호,,, 😂😂😂

bucketHaneul replied:
props를 spread로 전부 받게 되면 확장성을 더 높일 수 있다고 생각되면서도 너무 too much한 자율성을 주는 건 아닐까?? 잘못된 props가 알게모르게 넘어가지 않을까? 하는 생각이 조금 있었던 것 같습니다 ㅎㅎㅎ 🥺

jbee37142 commented on src/components/shared/IconButton/IconButton.tsx:
```diff
@@ -0,0 +1,18 @@
+import Styled from './IconButton.styles';
+
+interface IconButtonProps {
+  type?: 'button' | 'submit' | 'reset' | undefined;
```
optional과 undefined의 차이가 뭘까요?ㅎㅎ

bucketHaneul replied:
앗 optional는 자동적으로 `| undefined`를 타입에 추가해준다고 하는군용 😂😂😂

그렇다면 혹시 `| undefined` 대신 전부 optional로 변경하여도 괜찮을까요?? 아니면 optional대신 undefined를 작성하는게 더 나은 방법일 수도 있나요?? TS가 처음이라 아직은 타입 추론(?)이 어리둥절하네요오,,

jbee37142 replied:
앗, optional이 undefined로 infer되는 것은 맞지만 두 케이스를 사용하는 목적이 다릅니다!
음 좀 더 고민해보셔도 좋을 주제인데요, 아래 내용을 궁금하면 바로 보셔도 됩니다!

---

### Optional

```ts
interface Props {
  open?: boolean;
}
```

optional을 왜 사용했는지 돌이켜봤을 때, 넘기지 않아도 되는 즉 선택적으로 넘기도록 optional로 해줬을 것입니다.
이 Props를 같는 컴포넌트는 `open`을 넘기지 않아도 됩니다.

```tsx
<Component />
```

그리고 optional로 지정된 `open` props의 타입은 undefined로 추론이 됩니다.

### undefined

```ts
interface Props {
  open: boolean | undefined
}
```

이 때 open은 넘겨줘야 하는 props입니다!
`<Component />` // 에러가 발생!

이렇게 optional과 undefined type은 사용하는 목적이 다릅니다!ㅎㅎ

bucketHaneul replied:
오 이 부분을 이번에 코드를 짜면서 어리둥절 에러가 나서 알게 되었는데요!!!!! 😀😀😀

optional일 때는 넘겨도 되고, 안 넘겨도 된다! 대신 안넘기면 `undefined`로 추론이 된다!
optional이 아니면서 `| undefined`를 준다면, 무조건 넘기되 넘기는 props의 값이 undefined 일 수 있다!!

ㅎㅎㅎ 뭔가 머리로는 이해가 되는데, 아직 두 case에 해당하는 상황을 마주한 적이 없어서 조금 애매하네요!!!!! 호호 블로그에 정리하면서 다시 알아보고싶은 주제입니다 ㅎㅎㅎ 더 공부해보겠습니다~!

jbee37142 commented on src/components/shared/IconButton/IconButton.tsx:
```diff
@@ -0,0 +1,18 @@
+interface IconButtonProps {
+  type?: 'button' | 'submit' | 'reset' | undefined;
+  children: React.ReactNode;
+  onClick?: () => void;
```
button element의 onClick type과는 맞지 않는 typing인데 의도하신건가요?

bucketHaneul replied:
앗 아니요,,,!,,,, button element의 onClick type이 따로 존재하는지 몰랐습니다,,,!!!!
다음과 같이 수정했습니다!

```js
 onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
```

jbee37142 commented on src/components/shared/IconButton/IconButton.tsx:
```diff
@@ -0,0 +1,18 @@
+  onClick?: () => void;
+}
+
+const IconButton = ({ type, children, onClick }: IconButtonProps) => {
```
보아하니 다른 처리가 없는데 `Styled.Container` 그대로 export해도 아무 문제가 없을 것 같습니다.

jbee37142 commented on src/components/shared/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,35 @@
+
+interface DropdownProps {
+  labelText: string;
+  defaultOption: string;
```
HTML 표준과 다른 네이밍일 경우 이게 뭐지 하는 의심을 사게 됩니다.
input, select 모두 defaultValue라는 네이밍을 사용하고 있는데, defaultOption는 뭘까? 하는 의구심이 먼저 들 것 같아요!

jbee37142 replied:
defaultOption없이 사용하고 싶을 수도 있지 않을까요 ㅎㅎ

bucketHaneul replied:
HTML 표준과 다른 네이밍일 경우 충분히 헷갈릴 수 있겠네요

defaultOption을 defaultValue로 수정하고, optional 처리 하였습니다
감사합니다 !

jbee37142 commented on src/components/shared/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,35 @@
+    <Styled.Label>
+      {labelText}
+      <Styled.Select value={value} onChange={onSelect} required>
+        <option value="" selected disabled hidden>
+          {defaultOption}
+        </option>
```
이 `Dropdown`을 사용하는 입장에서 `options`로 전달한 option 말고 다른 option이 렌더링되면 당황스러울 것 같아요!

bucketHaneul replied:
앗 죄송하지만 제가 위 코멘트르 잘 이해하지 못 했습니다 ㅜ ㅜ 

`defaultValue`는 label과 같은 내용으로 select에서 제일 처음 보이는 option인데요! 지하철 역의 모음을 options로 전달할 때 `지하철 역`이라는 이름의 option을 disable selected 처리하여 넘겨주는 방법이 어색한 방법일까욤? 😮😮😮

jbee37142 replied:
음, 네! 엘라님 방식이 틀린 것은 절대 아니고, 제 생각에 조금 어색하다고 느낀 점을 좀 더 자세히 공유드려볼게요!

defaultValue와 options에 대해 생각을 해보면,
- Select에서 defaultValue 는 기본적으로 Select된 Value가 아닐까요? 👀 
- Select에 전달하는 options는 선택 가능한 value(그리고 label?)의 배열 아닐까요?👀 

두 가지가 어색하다고 생각했는데요,

1. 기본적으로 Select된 Value가 disabled인 것.
2. options로 전달하지 않은 value가 렌더링 된 것.

요런 이유로 defaultValue로 전달되는 값은 options에 포함되어 있어야 하지 않을까 하는 생각을 했습니다!ㅎㅎ

+) 근데 이제 보니 `hidden`처리가 되어있던데 placeholder 역할이 아닐지 의심해봅니다.

bucketHaneul replied:
아하!!!!!! 네 이해됐습니다!!

defaultValue는 options 중 하나이면서 이미 선택된 option이라고 기대할 수 있는데, 그렇지 않은 점이 어색하다고 느끼셨군요!!! 충분히 어색하다고 느낄 수 있는 부분이네요. 2단계 때 반영해보겠습니다 ❗❗❗

jbee37142 commented on src/components/shared/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import Styled from './Dropdown.styles';
+
+interface Option {
+  id: number;
+  value: string;
+}
```
value와 name이 다를 경우 이 Dropdown 컴포넌트는 사용할 수 없을 것 같아요 ㅎㅎ
(value에 id가 들어간다는 것을 내려보다가 알게 되었네요)

bucketHaneul replied:
하나의 option에서 user에게 보여지는 부분을 value로, 해당 옵션이 선택되었을 때 저장되는 값을 id로 하고 싶어서 위와 같은 DropDown 컴포넌트를 만들었는데요 ! 

위처럼 컴포넌트를 작성하고, 그에 맞춰 Dropdown 컴포넌트를 사용하면 된다고 생각했습니다! 너무 제약을 둔 것일까요?

jbee37142 replied:
물론 상황에 따라 다를테지만 보통 보여지는 부분은 label 실제 값으로 다루는 부분을 value라고 하긴 합니당 ㅋㅋㅋ
큰 상관은 없는데요, `id`가 number라서 저장되는 값이 `string`이면 사용 불가 판정 내려드립니다!ㅋㅋㅋ

jbee37142 commented on src/components/shared/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,35 @@
+}
+
+interface DropdownProps {
+  labelText: string;
```
labelText없이 사용하고 싶을 수도 있지 않을까요 ㅎㅎ

bucketHaneul replied:
ㅎ ㅎ 인정합니다! optional 처리하였습니다 >_<

jbee37142 commented on src/components/shared/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,35 @@
+  labelText: string;
+  defaultOption: string;
+  options: Option[];
+  value: string | number;
```
Option에선 value가 string 뿐이었어요. options로 렌더링되는 value랑 전달받는 value랑 다를 수도 있어서 buggy해보입니다!

jbee37142 commented on src/components/shared/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,35 @@
+          {defaultOption}
+        </option>
+        {options.map((option) => (
+          <option key={option.id} value={option.id} selected={option.value === defaultOption}>
```
`value`에 `id`가 전달될 것임을 내부 코드를 봐야만 알 수 있었네요!
`value`는 `value`로 전달되길 기대하지 않을까요?

jbee37142 replied:
selected는 전달받은 `defaultOption`이 아닌  `value`와 비교되어야 하지 않을까요?

bucketHaneul replied:
1. `option-tag`에 id가 아닌 value 그 자체가 전달되면, Dropdown를 사용할때마다 선택된 value를 가지고 id를 계속해서 찾아야합니다. 그래서 선택된 value에 대한 id를 넘겨주도록 하였는데요! 많이 어색한 방법일까요? ㅠ ㅠ

2.  앗,,,,맞습니다,,,! 수정하였습니다! 감사합니다 🙌🙌🙌

jbee37142 replied:
1. 아 의도하신대로 진행하는 것이 맞고 바로 위 코멘트와 동일한 코멘트였습니다 ㅎㅎ

jbee37142 commented on src/components/shared/Input/Input.tsx:
```diff
@@ -0,0 +1,47 @@
+      <Styled.Input
+        type={type}
+        value={value}
+        icon={icon}
```
`icon`이 렌더링되진 않고 padding-left만 조절을 하네요...? 👀 
Icon을 렌더링하는 것이 Input의 책임이 아니라고 생각하신 것 같은데, 그렇다면 이 padding-left 값은 Icon과 깊은 의존관계가 있으므로 이 네이밍을 받아서도 안 되고 padding-left를 이 Input이 처리하는 것이 아니라 외부에서 처리해줘야 할 것 같네요! ㅎㅎ

bucketHaneul replied:
해당 icon은 이번 프로젝트에서 input 안에 icon을 같이 주고 싶을 때, background-image로 들어가는 icon입니다!
따라서, icon을 받으면 backgrpund-image로 들어가고 그럼에 따라서 padding-left가 조절되고 있는 형태입니다!

jbee37142 replied:
네~ 코드에 그렇게 나와있어서 알고 있는데요 ㅎㅎ...
Input 컴포넌트의 역할일지 고민해보면 좋을 것 같습니다!
(덤으로 background-image로 Icon을 보여주는 것이 맞는지도 함께!!)

jbee37142 commented on src/components/shared/Input/Input.tsx:
```diff
@@ -0,0 +1,47 @@
+  return (
+    <Styled.Label>
+      {labelText}
+      <Styled.Input
```
input도 attirubutes를 전부 내려받으면 많아진 props를 조금이라도 해결할 수 있을 것 같습니다 ㅎㅎ

jbee37142 commented on src/components/shared/Input/Input.tsx:
```diff
@@ -0,0 +1,47 @@
+import Styled from './Input.styles';
+
+interface InputProps {
+  type: string;
```
기본 HTML input element는 default를 넣어주는데 이 Input은 props를 강제하고 있는 것이 아쉽네요 ㅜㅜ

bucketHaneul replied:
위 코멘트와 같이 해결하기 위해

```js
interface Props {
  labelText: string;
  icon?: string;
}

const Input = ({ labelText, icon,  ...props }: Props) => {
  return (
    <Styled.Label>
      {labelText}
      <Styled.Input  icon={icon} {...props} required />
    </Styled.Label>
  );
};
```

위처럼 코드를 수정하니, Input 컴포넌트를 사용하는 쪽에서 type이 추론되지 않는다고 뜹니다,,,,!,,,,,,,,,,,,,,,ㅜ ㅜ
혹시 안되는 이유를 알고 계신가요,,,?,,,

jbee37142 replied:
Input attribute를 Props에서 extends 해줘야 type 추론이 되지 않을까요 ㅎㅎ
엘라님이 정의해주신 Props는 labelText, icon 두개 뿐이라 추론될게 없을 것 같아요!!ㅎㅎㅎㅎㅎ

bucketHaneul replied:
호호 새로운 사실 또 알아가네요 >_< 이 부분도 도전해서 반영해보겠습니다 ㅎㅎㅎㅎㅎ

감사합니다 👍👍👍

jbee37142 commented on src/hooks/useFetch.tsx:
```diff
@@ -0,0 +1,32 @@
+import { API_METHOD, API_STATUS } from '../constants';
+
+const useFetch = (method: API_METHOD) => {
+  const [loading, setLoading] = useState<boolean>(false);
```
초기값으로 false를 전달해줬기 때문에 `loading`은 boolean임이 infer되어 `<boolean>`이렇게 명시적으로 generic을 전달할 필요가 없습니다 ㅎㅎ

bucketHaneul replied:
앗 그런가요???? 새로 알게된 사실이네요~! 감사합니다 >_<

jbee37142 commented on src/hooks/useFetch.tsx:
```diff
@@ -0,0 +1,32 @@
+        data,
+      });
+
+      return { status: API_STATUS.FULFILLED, data: response.data };
```
`status`라는 네이밍이 http status라고 예상되는데 별도로 정의해준 값이라서 혼란이 있지 않을까 생각됩니다 ㅎㅎ

bucketHaneul replied:
충분히 그럴 수 있겠네요!

 status -> state
APT_STATUS -> RESPONSE_STATE

변경하였습니다

jbee37142 commented on src/hooks/useFetch.tsx:
```diff
@@ -0,0 +1,32 @@
+        data,
+      });
+
+      return { status: API_STATUS.FULFILLED, data: response.data };
```
이렇게 api fetch 함수를 만들면 response로 전달되는 type들이 전부 infer되지 않을 것 같은데 잘 되던가요?

bucketHaneul replied:
흠 혹시 infer이 안되는 경우가 어떤 상황인지 알려주실 수 있을까요??? 😮😮😮
아무 문제 없이 코드가 돌아가긴 하는데요,,,!!!,,,,흑흑 TS가 처음이라 어떻게 봐야 추론이 안된다고 할 수 있는지 잘 모르겠습니다 ㅜ ㅜ

jbee37142 replied:
아 사용에 불편을 못느꼈으면 넘어가도 좋습니다!
`.`을 찍었을 때, autocomplete이 안 된다던가 cursor를 올렸을 때, type definition이 any (또는 unknown)으로 뜨면 추론이 안 되는 상황입니다 ㅎㅎ

jbee37142 commented on src/hooks/useNotify.tsx:
```diff
@@ -0,0 +1,26 @@
+    isVisible: false,
+  });
+
+  const Notification = () => (
```
보통 컴포넌트 내부에서 hooks를 호출하는데요,
이 경우 반대로 hooks에서 컴포넌트를 반환하고 있습니다.

이 useNodify hooks의 책임은 어떤 책임일까요?

bucketHaneul replied:
앗 제가 고민했던 부분입니다!!!! >_<

처음에는 Notification 컴포넌트는 가지지 않은체로 useNotify를 구현했었습니다 !
그런데 생각해보니 useNotify에 있는 하나의 state, 훅이 호출된 쪽에서 단 하나의 Notification 컴포넌트와만 관련됩니다.

그렇다면 useNotify라는 훅이 하나의 Notification 컴포넌트 그리고 이 컴포넌트와 관련된 하나의 state까지 가지면서 Notification 컴포넌트에 들어가는 message를 제어해도 되지 않을까?라고 생각하여 컴포넌트의 위치를 옮겼습니다!

jbee37142 reviewed (changes requested):
수고하셨어요 엘라님 ㅎㅎ
제가 이 프로젝트를 이어서 개발한다면? 이라는 입장에서 리뷰를 해보았습니다. (그러기엔 리뷰해야하는 코드 양이 너무나도 많네요...)

1. shared directory에 위치한 컴포넌트들은 재사용을 고려하여 만들어졌다는 가정하에 리뷰를 해봤습니다~
2. 오랜만에 리뷰어가 되어서 그동안 드리고 싶었던 리뷰를 못드려서 그랬는지 리뷰가 많아졌어요. (정신차리고 도중에 멈췄습니다)

여쭤보신 

> 비동기 통신에 대한 구조가 잘 잡혀있는지
기존 custom Hook에서 개선할 점이 있는지
custom Hook으로 더 분리할 수 있는 부분이 있는지
컴포넌트가 재사용성 있게 잘 구현되었는지

중, "컴포넌트 재사용성" 중심으로 리뷰가 되었고 (순서대로 하다보니)
비즈니스 로직 관련 (pages)는 거의 못봤어요ㅠㅠ

bucketHaneul committed:
refactor: IconButton과 TextButton 통합

- text props를 children으로 대체
- Buttontypes에 TRANSPARENT type 추가

bucketHaneul committed:
refactor: ButtonSize 타입에서 LARGE 삭제

- button width를 늘려주고 싶은 page 단에서 스타일 조절하도록 수정

bucketHaneul committed:
refactor: Modal 컴포넌트 리팩토링

- Modal 컴포넌트에서 closeButton 분리
- title 선택적으로 받을 수 있게 수정
- closebutton 선택적으로 받을 수 있게 수정
- ModalProps를 Props로 수정

bucketHaneul committed:
refactor: undefined를 가지는 type을 optional로 변경

bucketHaneul committed:
refactor: Button 컴포넌트 onClick 타입 수정

bucketHaneul committed:
refactor: Dropdown 컴포넌트 리팩토링

- defaultOption -> defaultValue로 수정
- defaultValue optional 처리

bucketHaneul committed:
refactor: 비동기 통신 응답 상태를 status에서 state로 네이밍 수정

bucketHaneul commented:
으학학 어마무시한 리뷰양에 깜짝 놀랐습니다!! 🙌🙌🙌
제이비님 덕분에 저도 다시 한 번 제가 놓친 부분과 실수한 부분들 다시 생각해 볼 수 있었습니다. 감사합니다 ~ !

최근에 컨디션이 안좋아서 리뷰 반영이 너무 늦어졌네요 ㅜㅜㅜㅜ 죄송합니다 ㅠㅜㅜㅜㅜ

bucketHaneul committed:
refactor: ButtonProps -> Props로 변경

jbee37142 reviewed (approved):
리뷰가 많았을텐데 전부 소화하신 것에 👏 👏 👏 
수고하셨습니다!ㅎㅎ 질문 주신 것에 대해선 코멘트 달아뒀습니다!
